java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,767 @@
+    Block afterLoop = currentBlock;
+    currentBlock = createBlock();
+    Block loopback = currentBlock;
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(afterLoop);
+    build(tree.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    currentBlock = createBranch(tree, currentBlock, afterLoop);
+    loopback.successors.add(currentBlock);
+    build(tree.variable());
+    build(tree.expression());
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildForStatement(ForStatementTree tree) {
+    Block falseBranch = currentBlock;
+    // process step
+    currentBlock = createBlock();
+    Block updateBlock = currentBlock;
+    for (StatementTree updateTree : Lists.reverse(tree.update())) {
+      build(updateTree);
+    }
+    continueTargets.addLast(currentBlock);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    breakTargets.addLast(falseBranch);
+    build(tree.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    Block body = currentBlock;
+    // process condition
+    ExpressionTree condition = tree.condition();
+    if (condition != null) {
+      currentBlock = createBranch(tree, body, falseBranch);
+      buildCondition(condition, body, falseBranch);
+    } else {
+      currentBlock = createUnconditionalJump(tree, body);
+    }
+    updateBlock.successors.add(currentBlock);
+    // process init
+    currentBlock = createBlock(currentBlock);
+    for (StatementTree init : Lists.reverse(tree.initializer())) {
+      build(init);
+    }
+  }
+
+  private void buildTryStatement(TryStatementTree tree) {
+    // FIXME only path with no failure constructed for now, (not taking try with resources into consideration).
+    TryStatementTree tryStatementTree = tree;
+    currentBlock = createBlock(currentBlock);
+    BlockTree finallyBlock = tryStatementTree.finallyBlock();
+    if (finallyBlock != null) {
+      build(finallyBlock);
+    }
+    currentBlock = createBlock(currentBlock);
+    build(tryStatementTree.block());
+    currentBlock = createBlock(currentBlock);
+    currentBlock.elements.add(tree);
+  }
+
+  private void buildThrowStatement(ThrowStatementTree tree) {
+    // FIXME this won't work if it is intended to be caught by a try statement.
+    ThrowStatementTree throwStatementTree = tree;
+    currentBlock = createUnconditionalJump(throwStatementTree, exitBlock);
+    build(throwStatementTree.expression());
+  }
+
+  private void buildSynchronizedStatement(SynchronizedStatementTree tree) {
+    SynchronizedStatementTree sst = tree;
+    // Naively build synchronized statement.
+    build(sst.block());
+    build(sst.expression());
+  }
+
+  private void buildUnaryExpression(UnaryExpressionTree tree) {
+    if(tree.is(Tree.Kind.PREFIX_DECREMENT, Tree.Kind.PREFIX_INCREMENT)) {
```

Comment:
@benzonico The fact that postfix operators return value before modification shouldn't change order in CFG, because loading of value should anyway happen before operation, i.e. order should be the same.